src/audio/mux/mux_generic.c

```diff
@@ -18,347 +18,528 @@
- */
-static void demux_s16le(struct audio_stream *sink,
-			const struct audio_stream *source, uint32_t frames,
-			struct mux_stream_data *data)
+static uint32_t mux_calc_frames_without_wrap_s16(struct comp_dev *dev,
+						 struct audio_stream *sink,
+						 const struct audio_stream
+						 **sources,
+						 struct mux_look_up *lookup)
 {
-	int32_t sample;
-	int16_t *dst;
-	uint32_t dst_idx;
-	uint8_t i;
-	uint8_t out_ch;
-
-	for (i = 0; i < frames; i++) {
-		for (out_ch = 0; out_ch < sink->channels; out_ch++) {
-			sample = calc_sample_s16le(source,
-						   i * source->channels,
-						   data->mask[out_ch]);
-
-			/* saturate to 16 bits */
-			dst_idx = i * sink->channels + out_ch;
-			dst = audio_stream_write_frag_s16(sink, dst_idx);
-			*dst = sat_int16(sample);
-		}
+	const struct audio_stream *source;
+	uint32_t frames;
+	uint32_t min_frames;
+	uint32_t elem;
+	void *ptr;
+
+	/* dest pointer for all copy_elems in lookup refers to the same
+	 * sink buffer (mux has one sink buffer), so dest min_frames
+	 * calculation based only on lookup table first element is sufficient.
+	 */
+	ptr = (int16_t *)lookup->copy_elem[0].dest -
+		lookup->copy_elem[0].out_ch;
+	min_frames = audio_stream_frames_without_wrap(sink, ptr);
+
+	for (elem = 0; elem < lookup->num_elems; elem++) {
+		source = sources[lookup->copy_elem[elem].stream_id];
+
+		ptr = (int16_t *)lookup->copy_elem[0].src -
+			lookup->copy_elem[0].in_ch;
+		frames = audio_stream_frames_without_wrap(source, ptr);
+
+		min_frames = (frames < min_frames) ? frames : min_frames;
 	}
+
+	return min_frames;
 }
 
-/* \brief Muxing 16 bit streams.
- *
- * Source streams are routed to sink with regard to routing bitmasks from
- * mux_stream_data structures array. Each source stream has bitmask for each
- * of it's channels describing to which channels of output stream it
- * contributes.
- *
- * \param[in,out] sink Destination buffer.
- * \param[in,out] sources Array of source buffers.
- * \param[in] frames Number of frames to process.
- * \param[in] data Array of parameters describing channel count and routing for
- *		   each stream.
- */
-static void mux_s16le(struct audio_stream *sink,
-		      const struct audio_stream **sources, uint32_t frames,
-		      struct mux_stream_data *data)
+static uint32_t demux_calc_frames_without_wrap_s32(struct comp_dev *dev,
+						   struct audio_stream *sink,
+						   const struct audio_stream
+						   *source,
+						   struct mux_look_up *lookup)
+{
+	uint32_t frames;
+	uint32_t min_frames;
+	void *ptr;
+
+	/* for demux we process each source buffer separately - dest/src for
+	 * each copy_elem refers to the same sink/source buffer, so min_frames
+	 * calculation based only on lookup table first element is sufficient.
+	 */
+	ptr = (int32_t *)lookup->copy_elem[0].dest -
+		lookup->copy_elem[0].out_ch;
+	min_frames = audio_stream_frames_without_wrap(sink, ptr);
+
+	ptr = (int32_t *)lookup->copy_elem[0].src -
+		lookup->copy_elem[0].in_ch;
```

Comment:
and here too

Reply:
Here, it's indended. Please see comment in above it (lines: 129-132):
```
/* for demux we process each source buffer separately - dest/src for
 * each copy_elem refers to the same sink/source buffer, so min_frames
* calculation based only on lookup table first element is sufficient.
*/
```